p2p/tools/network.py

```diff
@@ -0,0 +1,189 @@
+import asyncio
+import logging
+import random
+from typing import (
+    Dict,
+    Iterable,
+    NamedTuple,
+    Tuple,
+)
+
+
+class Address(NamedTuple):
+    transport: str
+    host: str
+    port: int
+
+    def __repr__(self):
+        return '<%s %s>' % (self.__class__.__name__, self)
+
+    def __str__(self):
+        return '%s:%s:%s' % (self.transport, self.host, self.port)
+
+
+class MemoryTransport(asyncio.Transport):
+    """
+    Direct connection between a StreamWriter and StreamReader.
+    """
+
+    def __init__(self, address: Address, reader: asyncio.StreamReader) -> None:
+        super().__init__()
+        self._address = address
+        self._reader = reader
+
+    def get_extra_info(self, name, default=None):
+        if name == 'peername':
+            return (self._address.host, self._address.port)
+        else:
+            return super().get_extra_info(name, default)
+
+    def write(self, data: bytes) -> None:
+        self._reader.feed_data(data)
+
+    def writelines(self, data: Iterable[bytes]) -> None:
+        for line in data:
+            self._reader.feed_data(line)
+            self._reader.feed_data(b'\n')
+
+    def write_eof(self) -> None:
+        self._reader.feed_eof()
+
+    def can_write_eof(self) -> bool:
+        return True
+
+    def is_closing(self) -> bool:
+        return False
+
+    def close(self) -> None:
+        self.write_eof()
+
+
+def mempipe(address) -> Tuple[asyncio.StreamReader, asyncio.StreamWriter]:
+    """In-memory pipe, returns a ``(reader, writer)`` pair.
+
+    .. versionadded:: 0.1
+
+    """
+
+    reader = asyncio.StreamReader()
+
+    # TODO: Calls to `writer.drain` will block until the corresponding data has
+    # actually been read by the reader.  This is not how things actually behave
+    # in a real networked environment as the call to `writer.drain` will return
+    # once the data has been sent over the protocol, so...
+
+    # Preliminary investigation suggests that this must be done in the
+    # `Protocol` class
```

Comment:
@gsalgado I'd like to get your thoughts on this. My current general thoughts on how to approach this are:

Setup two `reader,writer` pairs, `r1,w1` and `r2,w2` such that `w2` is the main writer and `r1` is the main reader.  Data flows `w2 -> r2 -> w1 -> r1`.  Then we create a long running asynchronous loop which continually reads from `r2` and writes what it reads to `w1`.

You can think of this as buffering the data from the main writer to the main reader via an intermediate step (which sort of unintentionally simulates the actual networking).

Reply:
That sounds ok to me, but I worry a bit that all the extra complexity added by the in-memory mock network will make it even harder to debug failing tests. Some of our tests are already pretty hard to debug because they are implemented as integration tests that rely on multiple components connecting (sometimes over the network) and talking to each other, but I'm not sure simply removing network access from the equation is the right solution -- I think it'd be better to try and rewrite those as unit tests so that they setup/exercise just the pieces they want to test. I understand this is usually tricky and not the most exciting work, so don't let that deter you from working on this mock-networking approach, but I thought it was worth bringing it up